tsfc/loopy.py

```diff
@@ -181,15 +224,44 @@ def statement_evaluate(leaf, ctx):
     expr = leaf.expression
     if isinstance(expr, gem.ListTensor):
         ops = []
-        var = ctx.pymbolic_variable(expr)
-        index = ()
-        if isinstance(var, p.Subscript):
-            var, index = var.aggregate, var.index_tuple
+        var, index = ctx.pymbolic_variable_and_destruct(expr)
         for multiindex, value in numpy.ndenumerate(expr.array):
             ops.append(lp.Assignment(p.Subscript(var, index + multiindex), expression(value, ctx), within_inames=ctx.active_inames()))
         return ops
     elif isinstance(expr, gem.Constant):
         return []
+    elif isinstance(expr, gem.ComponentTensor):
+        statements = []
+        idx = ctx.pymbolic_indices(expr)
+        var, sub_idx = ctx.pymbolic_variable_and_destruct(expr)
+        lhs = p.Subscript(var, idx + sub_idx)
+        with active_indices(expr.multiindex, idx, ctx) as ctx_active:
+            statements = [lp.Assignment(lhs, expression(expr.children[0], ctx_active), within_inames=ctx_active.active_inames())]
+        return statements
+    elif isinstance(expr, gem.Inverse):
+        idx = ctx.pymbolic_indices(expr)
+        var = ctx.pymbolic_variable(expr)
+        lhs = (SubArrayRef(idx, p.Subscript(var, idx)),)
+
+        idx_reads = ctx.pymbolic_indices(expr.children[0])
+        var_reads = ctx.pymbolic_variable(expr.children[0])
+        reads = (SubArrayRef(idx_reads, p.Subscript(var_reads, idx_reads)),)
+        rhs = p.Call(p.Variable("inv"), reads)
+
+        return [lp.CallInstruction(lhs, rhs, within_inames=ctx.active_inames())]
+    elif isinstance(expr, gem.Solve):
+        idx = ctx.pymbolic_indices(expr)
+        var = ctx.pymbolic_variable(expr)
+        lhs = (SubArrayRef(idx, p.Subscript(var, idx)),)
+
+        reads = ()
+        for child in expr.children:
+            idx_reads = ctx.pymbolic_indices(child)
+            var_reads = ctx.pymbolic_variable(child)
+            reads += (SubArrayRef(idx_reads, p.Subscript(var_reads, idx_reads)),)
```

Comment:
For collecting in a list, I prefer making a list and appending. You can cast to tuple after the loop.